examples/algstruct.ctt

```diff
@@ -0,0 +1,671 @@
+  : prop (hasRightInverse G op id inv)
+  = let
+    B (x : G) : U
+      = Path G (op x (inv x)) id
+    h (x : G) : prop (B x)
+      = sG (op x (inv x)) id
+  in propPi G B h
+
+hasInverse (G : U) (op : G -> G -> G) (id : G) (inv : G -> G) : U
+  = (_ : hasLeftInverse G op id inv)
+  * (hasRightInverse G op id inv)
+
+prophasInverse (G : U) (sG : set G) (op : G -> G -> G) (id : G) (inv : G -> G)
+  : prop (hasInverse G op id inv)
+  = propAnd (hasLeftInverse G op id inv) (hasRightInverse G op id inv)
+      (prophasLeftInverse G sG op id inv) (prophasRightInverse G sG op id inv)
+
+-- Distributivity
+
+isLeftDistributive (R : U) (add : R -> R -> R) (mul : R -> R -> R) : U
+  = (a b c : R) -> Path R (mul a (add b c)) (add (mul a b) (mul a c))
+
+propisLeftDistributive (R : U) (sR : set R) (add : R -> R -> R) (mul : R -> R -> R)
+  : prop (isLeftDistributive R add mul)
+  = let
+    B (a b c : R) : U
+      = Path R (mul a (add b c)) (add (mul a b) (mul a c))
+    h (a b c : R) : prop (B a b c)
+      = sR (mul a (add b c)) (add (mul a b) (mul a c))
+  in propPi3 R B h
+
+isRightDistributive (R : U) (add : R -> R -> R) (mul : R -> R -> R) : U
+  = (a b c : R) -> Path R (mul (add b c) a) (add (mul b a) (mul c a))
+
+propisRightDistributive (R : U) (sR : set R) (add : R -> R -> R) (mul : R -> R -> R)
+  : prop (isRightDistributive R add mul)
+  = let
+    B (a b c : R) : U
+      = Path R (mul (add b c) a) (add (mul b a) (mul c a))
+    h (a b c : R) : prop (B a b c)
+      = sR (mul (add b c) a) (add (mul b a) (mul c a))
+  in propPi3 R B h
+
+isDistributive (R : U) (add : R -> R -> R) (mul : R -> R -> R) : U
+  = (_ : isLeftDistributive R add mul)
+  * (isRightDistributive R add mul)
+
+propisDistributive (R : U) (sR : set R) (add : R -> R -> R) (mul : R -> R -> R)
+  : prop (isDistributive R add mul)
+  = propAnd (isLeftDistributive R add mul) (isRightDistributive R add mul)
+      (propisLeftDistributive R sR add mul) (propisRightDistributive R sR add mul)
+
+-- Structure preserving properties of maps
+
+preservesOp (A B : U) (oA : A -> A -> A) (oB : B -> B -> B) (f : A -> B) : U
+  = (a0 a1 : A) -> Path B (f (oA a0 a1)) (oB (f a0) (f a1))
+
+proppreservesOp (A B : U) (sB : set B) (oA : A -> A -> A) (oB : B -> B -> B)
+    (f : A -> B) : prop (preservesOp A B oA oB f)
+  = propPi2 A (\ (a0 a1 : A) -> Path B (f (oA a0 a1)) (oB (f a0) (f a1)))
+      (\ (a0 a1 : A) -> sB (f (oA a0 a1)) (oB (f a0) (f a1)))
+
+preservesId (A B : U) (iA : A) (iB : B) (f : A -> B) : U
+  = Path B (f iA) iB
+
+proppreservesId (A B : U) (sB : set B) (iA : A) (iB : B) (f : A -> B)
+  : prop (preservesId A B iA iB f)
+  = sB (f iA) iB
+
+
+-- Algebraic structures
+--------------------------------------------------------------------------------
+
+-- Group-like structures
+------------------------
+
+-- Monoid
+
+ismonoid (M : SET) : U
+  = (op : M.1 -> M.1 -> M.1)
+  * (_ : isAssociative M.1 op)
+  * (id : M.1)
+  * (hasIdentity M.1 op id)
+
```

Comment:
The code would be easier to read if you had explicit accessor functions for the structures. This way you can write for example `monoidId m` instead of `m.2.2.1`.